README.md

```diff
@@ -1 +1,420 @@
+>
+> — [DAG-JSON Spec][DAG-JSON]
+
+As it is critical for guarding against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+### 2.1.1 Caching & Invalidation
+
+Signing CIDs has two additional caching effects
+
+* Signing CIDs enables a simple strategy for caching validation by CID
+* Such a strategy also requires accounting for revocation of the signing keys, and so need to be tagged with this additional metadata
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `%x55`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+After being decoded from [unsigned varint]s, a varsig includes the following segments:
+
+```abnf
+varsig = multibase-prefix %x34 varsig-header varsig-body
+multibase-prefix = ALPHA ; Multibase
+varsig-header = unsigned-varint ; Usually the public key code from Multicodec
+varsig-body = *OCTET; Zero or more segments required by the kind of varsig (e.g. raw bytes, hash algorithm, etc)
+```
+
+For example, here is an EdDSA signature for some content encoded as DAG-PB:
+
+`%x34ed01ae3784f03f9ee1163382fa6efa73b0c31ecf58c899c836709303ba4621d1e6df20e09aaa568914290b7ea124f5b38e70b9b69c7de0d216880eac885edd41c302`
+
+### 3.1 Varsig Prefix
+
+The varsig prefix MUST be `%x34`.
+
+### 3.2 Signature Header
+
+The prefix of the signature algorithm. This is often the [multicodec] of the associated public key, but MAY be unique for the signature type. The code MAY live outside the multicodec table. This field MUST act as a discriminant for how many expected fields come in the varsig body, and what each of them mean.
+
+### 3.3 Varsig Body
+
+The varsig body MUST consist of one or more segments, and MUST be defined by the signature algorithm.
+
+Some examples include:
+
+* Raw signature bytes only
+* CID of [DKIM] certification transparency record, and raw signature bytes
+* Hash algorithm multicodec prefix, data encoding prefix, signature counter, nonce, HMAC, and raw signature bytes
+
+# 4 Payload Encoding
+
+The [IPLD] data model is encoding agnostic by design. This is very convenient in many applications, such as making for very convenient conversions between types for transmission versus encoding. Unfortunately signatures require signing over specific bytes, and thus over a specific encoding of the data.
+
+To facilitate this, the type `encoding-info` MAY be used:
+
+``` abnf
+encoding-info
+  = %x55   ; Raw bytes multicodec prefix
+  / %x70   ; DAG-PB multicodec prefix
+  / %x71   ; DAG-CBOR multicodec prefix
+  / %x0129 ; DAG-JSON multicodec prefix
+  / %x6A77 ; JWT
```

Comment:
Hmm interesting; can you tell me more? Is that to normalize the fields across CACAO types, or so that you know which fields to hoist from a flat structure into the header, or something else?

Reply:
> so that you know which fields to hoist from a flat structure into the header

This primarily.

As for the DIDs, in UCAN they are strings, for CACAO they need to be multidid encoded.

Reply:
Yeah that makes sense. Thinking about this, we're liekly going to need a bunch of these for a bunch of cases. IIRC a CACAO itself is encoding agnostic by design. How would you feel about something like this?

```abnf
  / %xCAC0 cacao-encoding

cacao-encoding
  = %x3C ; ReCap
  / %x51 ; SIWE
  / %xCA ; UCAN
```

# Aside

I also vaguely remember discussing there being a difference between the CACAO version and the version of the token that it maps to, but I don't see that in the latest version of https://github.com/ChainAgnostic/CAIPs/pull/196 — maybe not required, but I thought I'd check.

Reply:
I don't think we need to add CACAO specific encoding-info to this spec. These can be defined in the CACAOv3 CAIP only I think?

Old versions of CACAO doesn't use varsig, so doesn't really make sense to specify codes for them.

Reply:
Hmm, okay. Let's say for example that you cited where UCAN has `ucv` in the header. `ucv` is not a standard JWT field, but we do need to sign over the JWT-serialized format. Would the JWT serializer know that a CACAO `v` should be serialized as `ucv` in the header (I mean, we can also move this field to the claims/body, but it is basically header info since it describes the shape of the body).

Reply:
Yes exactly, the JWT serializer would need to know this. Moving the field I think could be a good option unless we want to specify a custom UCAN codec.